File: BinaryTrees/zigzag_level_order.py
from collections import deque
from typing import List, Optional

from tree_node import TreeNode


class ZigzagLevelOrder:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        level_orders = []

        if not root:
            return level_orders

        queue = deque([root])
        left_to_right = True

        while queue:
            order = deque()
            level_size = len(queue)
            for _ in range(level_size):
                node = queue.popleft()
                if left_to_right:
                    order.append(node.val)
                else:
                    order.appendleft(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            level_orders.append(list(order))
            left_to_right = not left_to_right

        return level_orders


# Time Complexity: O(n) where n is the number of nodes in the tree, as we visit each node once.
# Space Complexity: O(w) where w is the maximum width of the tree,
# which in the worst case (a complete binary tree) is O(n/2) = O(n).
# In the best case of a skewed tree, the width would be O(1), leading to O(1) space complexity.

File: BinaryTrees/zigzagLevelOrder.js
/*Given the root of a binary tree, return the zigzag level order traversal of its nodes' values. 
(i.e., from left to right, then right to left for the next level and alternate between).*/

const zigzagLevelOrder = (root) => {
    const treeTraversalOrder = [];

    if (!root) {
        return treeTraversalOrder;
    }

    let direction = 0;

    let queue = [root];

    while (queue.length) {
        const nextLevelQueue = [];
        const levelTraversalOrder = [];

        for (let node of queue) {
            if (direction % 2 == 0) { //right to left
                levelTraversalOrder.push(node.val)
            } else { //left to right
                levelTraversalOrder.unshift(node.val);
            }

            if (node.left) nextLevelQueue.push(node.left);
            if (node.right) nextLevelQueue.push(node.right);
        }

        treeTraversalOrder.push(levelTraversalOrder);
        queue = nextLevelQueue;
        direction++; //alternate directions
    }

    return treeTraversalOrder;
}